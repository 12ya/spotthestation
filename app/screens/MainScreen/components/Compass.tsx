/* eslint-disable react-native/no-inline-styles */
import React, { useState, useEffect } from 'react'
import { View, Text, Image, ViewStyle, ImageStyle, TextStyle } from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import CompassHeading from 'react-native-compass-heading'
import { Icon } from '../../../components'
import { colors } from '../../../theme'
const compassLine = require('../../../../assets/icons/compass-line.png')

const LINE_LENGTH = 260
const LETTER_OFFSET = 10

const normalizeHeading = (heading: number) => {
  if (heading < 0) return 360 + heading
  if (heading > 360) return heading - 360
  return heading
}

const isInHeadingRange = (left, right, heading) => {
  if (right < left) return heading > left || heading < right
  return heading >= left && heading <= right
}

const headingOffset = (h1, h2) => {
  if (h2 < h1) return h2 - (h1 - 360)
  return h2 - h1
}

export const Compass = ({ issPosition, isFullScreen }) => {
  const topInset = useSafeAreaInsets().top
  const [heading, setHeading] = useState(0)

  const left = normalizeHeading(heading - 45)
  const right = normalizeHeading(heading + 45)

  const letterPositions = [
    { letter: 'N', offset: 0 },
    { letter: 'E', offset: 90 },
    { letter: 'S', offset: 180 },
    { letter: 'W', offset: 270 },
  ].filter(letter => isInHeadingRange(left, right, letter.offset))

  useEffect(() => {
    CompassHeading.start(1, (result) => {
      setHeading(Number(result.heading))
    }).catch((err) => {
      console.log(err)
    })

    return () => {
      CompassHeading.stop().catch((err) => {
        console.log(err)
      })
    }
  }, [])

  const issVisible = isInHeadingRange(left, right, issPosition)

  return (
    <View style={[$container, { marginTop: isFullScreen ? topInset + 24 : 0 }]}>
      <Image source={compassLine} style={$line} />
      {issVisible && <Icon icon="iss" size={24} containerStyle={[$issIcon, { marginLeft: (headingOffset(left, issPosition) / 90) * LINE_LENGTH }]} />}
      <View style={$lettersContainer}>
        {letterPositions.map(({ letter, offset }) => {
          const letterPosition = (headingOffset(left, offset) / 90) * LINE_LENGTH
          return (
            <View key={letter} style={[$letterContainer, { marginLeft: letterPosition }]}>
              <Text
                key={letter}
                style={$letter}
              >
                {letter}
              </Text>
              <View style={$verticalLine} />
            </View>
          )
        })}
      </View>
    </View>
  )
}

const $container: ViewStyle = {
  height: 20,
  position: 'absolute',
  top: 0,
  alignSelf: 'center',
  width: LINE_LENGTH
}

const $issIcon: ViewStyle = {
  bottom: -10,
  position: 'absolute'
}

const $letter: TextStyle = {
  color: colors.palette.neutral100,
  marginBottom: 5,
  textAlign: 'left'
}

const $letterContainer: ViewStyle = {
  alignItems: 'center',
  marginBottom: 2
}

const $lettersContainer: ViewStyle = {
  bottom: -10,
  flexDirection: 'row',
  height: 'auto',
  left: 0,
  marginTop: -LETTER_OFFSET
}

const $line: ImageStyle = {
  bottom: 0,
  left: 0,
  position: 'absolute',
  width: LINE_LENGTH,
}

const $verticalLine: ViewStyle = {
  borderColor: colors.palette.neutral100,
  borderWidth: 1,
  height: 6,
  width: 1
}
