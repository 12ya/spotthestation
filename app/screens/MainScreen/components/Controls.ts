import { Camera, EventDispatcher, Spherical, TOUCH, Vector2, Vector3, BaseEvent } from "three"
import { NativeTouchEvent } from "react-native"

const STATE = {
  NONE: -1,
  TOUCH_ROTATE: 3,
}

export class Controls extends EventDispatcher {
  object: Camera & {
    fov: number
    top: number
    right: number
    left: number
    bottom: number
    zoom: number
    updateProjectionMatrix: () => void
    isOrthographicCamera?: boolean
    isPerspectiveCamera?: boolean
  }

  enabled: boolean
  target: Vector3
  minDistance: number
  maxDistance: number
  minPolarAngle: number
  maxPolarAngle: number
  minAzimuthAngle: number
  maxAzimuthAngle: number
  enableRotate: boolean
  rotateSpeed: number
  touches: { ONE: TOUCH }
  target0: Vector3
  position0: Vector3
  update: () => boolean
  private changeEvent: BaseEvent
  private startEvent: BaseEvent
  private endEvent: BaseEvent
  private state
  private spherical: Spherical
  private sphericalDelta: Spherical
  private scale
  private panOffset: Vector3
  private rotateStart: Vector2
  private rotateEnd: Vector2
  private rotateDelta: Vector2
  saveState: () => void
  reset: () => void
  private rotateLeft: (angle: number) => void
  private rotateUp: (angle: number) => void
  width: number
  getElementWidth: () => number
  height: number
  getElementHeight: () => number
  private handleTouchStartRotate: (event: NativeTouchEvent) => void
  private handleTouchMoveRotate: (event: NativeTouchEvent) => void
  onTouchStart: (event: NativeTouchEvent) => void
  onTouchMove: (event: NativeTouchEvent) => void
  onTouchEnd: (event: NativeTouchEvent) => void

  constructor(object) {
    super()
    this.object = object
    this.enabled = true
    this.target = new Vector3()
    this.minDistance = 0
    this.maxDistance = Infinity
    this.minPolarAngle = 0 // radians
    this.maxPolarAngle = Math.PI // radians
    this.minAzimuthAngle = -Infinity // radians
    this.maxAzimuthAngle = Infinity // radians
    this.enableRotate = true
    this.rotateSpeed = 0.5
    this.touches = { ONE: TOUCH.ROTATE }
    this.changeEvent = { type: "change" }
    this.startEvent = { type: "start" }
    this.endEvent = { type: "end" }
    this.state = STATE.NONE
    this.spherical = new Spherical()
    this.sphericalDelta = new Spherical()
    this.scale = 1
    this.panOffset = new Vector3()
    this.rotateStart = new Vector2()
    this.rotateEnd = new Vector2()
    this.rotateDelta = new Vector2()

    this.saveState = () => {
      this.target0.copy(this.target)
      this.position0.copy(this.object.position)
    }

    this.reset = () => {
      this.target.copy(this.target0)
      this.object.position.copy(this.position0)
      this.object.updateProjectionMatrix()
      this.dispatchEvent(this.changeEvent)
      this.update()
      this.state = STATE.NONE
    }

    this.rotateLeft = (angle) => {
      this.sphericalDelta.theta -= angle
    }

    this.rotateUp = (angle) => {
      this.sphericalDelta.phi -= angle
    }

    this.width = 0
    this.getElementWidth = () => {
      return this.width
    }
    this.height = 0
    this.getElementHeight = () => {
      return this.height
    }

    this.handleTouchStartRotate = ({ touches }) => {
      if (touches.length === 1) {
        this.rotateStart.set(touches[0].pageX, touches[0].pageY)
      } else {
        const x = 0.5 * (touches[0].pageX + touches[1].pageX)
        const y = 0.5 * (touches[0].pageY + touches[1].pageY)
        this.rotateStart.set(x, y)
      }
    }

    this.handleTouchMoveRotate = ({ touches }) => {
      if (touches.length === 1) {
        this.rotateEnd.set(touches[0].pageX, touches[0].pageY)
      } else {
        const x = 0.5 * (touches[0].pageX + touches[1].pageX)
        const y = 0.5 * (touches[0].pageY + touches[1].pageY)
        this.rotateEnd.set(x, y)
      }
      this.rotateDelta.subVectors(this.rotateEnd, this.rotateStart).multiplyScalar(this.rotateSpeed)
      this.rotateLeft((2 * Math.PI * this.rotateDelta.x) / this.getElementHeight())
      this.rotateUp((2 * Math.PI * this.rotateDelta.y) / this.getElementHeight())
      this.rotateStart.copy(this.rotateEnd)
    }

    this.onTouchStart = (event) => {
      if (this.enabled === false) return

      switch (event.touches.length) {
        case 1:
          switch (this.touches.ONE) {
            case TOUCH.ROTATE:
              if (this.enableRotate === false) return
              this.handleTouchStartRotate(event)
              this.state = STATE.TOUCH_ROTATE
              break
            default:
              this.state = STATE.NONE
          }
          break
        default:
          this.state = STATE.NONE
      }
      if (this.state !== STATE.NONE) {
        this.dispatchEvent(this.startEvent)
      }
    }

    this.onTouchMove = (event) => {
      if (this.enabled === false) return

      switch (this.state) {
        case STATE.TOUCH_ROTATE:
          if (this.enableRotate === false) return
          this.handleTouchMoveRotate(event)
          this.update()
          break
        default:
          this.state = STATE.NONE
      }
    }

    this.onTouchEnd = () => {
      if (this.enabled === false) return
      this.dispatchEvent(this.endEvent)
      this.state = STATE.NONE
    }

    this.target0 = this.target.clone()
    this.position0 = this.object.position.clone()

    this.update = (() => {
      const offset = new Vector3()
      return () => {
        const position = this.object.position
        offset.copy(position).sub(this.target)

        this.spherical.setFromVector3(offset)

        this.spherical.theta += this.sphericalDelta.theta
        this.spherical.phi += this.sphericalDelta.phi

        this.spherical.theta = Math.max(
          this.minAzimuthAngle,
          Math.min(this.maxAzimuthAngle, this.spherical.theta),
        )
        this.spherical.phi = Math.max(
          this.minPolarAngle,
          Math.min(this.maxPolarAngle, this.spherical.phi),
        )
        this.spherical.makeSafe()
        this.spherical.radius *= this.scale
        this.spherical.radius = Math.max(
          this.minDistance,
          Math.min(this.maxDistance, this.spherical.radius),
        )
        this.target.add(this.panOffset)

        offset.setFromSpherical(this.spherical)

        position.copy(this.target).add(offset)
        this.object.lookAt(this.target)
        this.sphericalDelta.set(0, 0, 0)
        this.panOffset.set(0, 0, 0)
        this.scale = 1

        return false
      }
    })()

    this.update()
  }
}
