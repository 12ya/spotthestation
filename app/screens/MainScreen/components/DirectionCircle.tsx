import { StyleFn, useStyles } from "../../../utils/useStyles"
/* eslint-disable @typescript-eslint/no-unsafe-call */
import React, { useMemo } from "react"
import { ViewStyle, View, useWindowDimensions, PixelRatio } from "react-native"
import Svg, { Path } from "react-native-svg"
import { arc as d3Arc } from "d3-shape"
import { degToRad } from "../../../utils/geometry"

const arc = d3Arc()

export const DirectionCircle = ({ screenX, screenY, setIsSpotted }) => {
  const { $container, $innerCircle, $outerCircle } = useStyles(styles)

  const dimensions = useWindowDimensions()
  const width = useMemo(() => dimensions.width * PixelRatio.get(), [dimensions])
  const height = useMemo(() => dimensions.height * PixelRatio.get(), [dimensions])

  const x = useMemo(() => screenX - width / 2, [screenX, width])
  const y = useMemo(() => screenY - height / 2, [screenY, height])
  const arrowAngle = useMemo(() => Math.atan2(x, -y), [x, y])
  const alignDistance = useMemo(() => Math.abs(x), [x])
  const viewDistance = useMemo(() => Math.max(0, alignDistance - width / 2), [alignDistance, width])
  const distance = useMemo(() => Math.sqrt(x * x + y * y), [x, y])
  const isSpotted = distance <= 70

  setIsSpotted(isSpotted)

  const outerArc = useMemo(() => {
    return arc({
      outerRadius: 45,
      innerRadius: 42,
      startAngle: 0,
      endAngle: degToRad(360),
    })
  }, [])

  const viewArc = useMemo(() => {
    const spread = degToRad(Math.max(0, 180 - Math.floor(viewDistance / 4)))
    return arc({
      outerRadius: 45,
      innerRadius: 42,
      startAngle: arrowAngle - spread,
      endAngle: arrowAngle + spread,
    })
  }, [viewDistance, arrowAngle])

  const alignArc = useMemo(() => {
    const spread =
      alignDistance < 20
        ? 180
        : degToRad(Math.min(180, Math.max(0, 210 - Math.floor(alignDistance / 2))))
    return arc({
      outerRadius: 45,
      innerRadius: isSpotted ? 30 : 42,
      startAngle: arrowAngle - spread,
      endAngle: arrowAngle + spread,
    })
  }, [alignDistance, arrowAngle, isSpotted])

  return (
    <View style={$container}>
      {distance > 150 && (
        <Svg
          width={22}
          height={24}
          transform={[{ rotate: `${arrowAngle}rad` }, { translateY: -65 }]}
        >
          <Path
            d="M 10.436 1.683 L 19.923 21.895 C 19.965 22.01 19.974 22.135 19.948 22.255 C 19.922 22.375 19.862 22.485 19.775 22.572 C 19.688 22.659 19.578 22.719 19.458 22.745 C 19.338 22.772 19.213 22.763 19.098 22.72 L 10.436 19.833 L 1.774 22.72 C 1.658 22.763 1.533 22.772 1.413 22.745 C 1.293 22.719 1.183 22.659 1.096 22.572 C 1.009 22.485 0.949 22.375 0.923 22.255 C 0.897 22.135 0.906 22.01 0.949 21.895 L 10.436 1.683 Z"
            fill={viewDistance === 0 ? "#11CE69" : "#fff"}
            stroke={viewDistance === 0 ? "#11CE69" : "#fff"}
            strokeWidth="1.5"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </Svg>
      )}
      <Svg style={$outerCircle} width={150} height={150}>
        <Path x={75} y={75} d={outerArc} fill="#ffffff44" />
        <Path x={75} y={75} d={viewArc} fill="#fff" />
        {viewDistance === 0 && <Path x={75} y={75} d={alignArc} fill={"#11CE69"} />}
      </Svg>
      {!isSpotted && <View style={$innerCircle}></View>}
    </View>
  )
}

const styles: StyleFn = ({ scale }) => {
  const $container: ViewStyle = {
    position: "absolute",
    width: "100%",
    height: "100%",
    alignItems: "center",
    justifyContent: "center",
  }

  const $innerCircle: ViewStyle = {
    width: scale(50),
    height: scale(50),
    borderRadius: scale(50),
    position: "absolute",
    alignSelf: "center",
    backgroundColor: "#fff",
  }

  const $outerCircle: ViewStyle = {
    position: "absolute",
    alignSelf: "center",
    width: scale(70),
    height: scale(70),
  }

  return { $container, $innerCircle, $outerCircle }
}
