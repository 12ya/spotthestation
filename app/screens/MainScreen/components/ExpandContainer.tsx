import React, { ReactNode, useState } from "react"
import { ViewStyle, View, TextStyle, ImageStyle } from "react-native"
import { Text, Icon } from "../../../components"
import { TxKeyPath } from "../../../i18n"
import { typography, colors } from "../../../theme"

export interface ExpandContainerProps {
  defaultValue?: boolean
  itemsCount?: number
  title: TxKeyPath
  expandble?: boolean
  children?: ReactNode
}

export function ExpandContainer({ defaultValue = true, title, expandble = true, children, itemsCount }: ExpandContainerProps) {
  const [expanded, setExpanded] = useState(defaultValue)
  return (
    <View style={$container}>
      <View style={$headContainer}>
        <View style={$titleContainer}>
          <Text tx={title} style={$title} />
          {itemsCount > 1 && <Text text={` (${itemsCount})`} style={$title} />} 
        </View>
        {expandble && <Icon icon="chevronDown" size={18} onPress={() => setExpanded(!expanded)} style={expanded && $up} />}
      </View>
      {expanded && children}
    </View>
  )
}

const $container: ViewStyle = {
  width: "100%",
  marginTop: 36
}

const $headContainer: ViewStyle = {
  marginBottom: 18,
  flexDirection: "row",
  justifyContent: "space-between"
}

const $titleContainer: ViewStyle = {
  flexDirection: "row",
}

const $title: TextStyle = {
  fontFamily: typography.primary?.normal,
  fontSize: 13,
  lineHeight: 16,
  color: colors.palette.neutral450,
  textTransform: "uppercase"
}

const $up: ImageStyle = {
  transform: [{ rotate: "180deg" }]
}
