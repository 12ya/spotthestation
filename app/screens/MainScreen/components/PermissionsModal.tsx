import { StyleFn, useStyles } from "../../../utils/useStyles"
import React from "react"
import { ViewStyle, View, TextStyle, PressableProps } from "react-native"
import { Icon, Text, Button } from "../../../components"
import { typography } from "../../../theme"
import { colors } from "../../../theme/colors"
import { normalizeHeight } from "../../../utils/normalizeHeight"

export interface PermissionsModalProps {
  body: string
  onClose?: PressableProps["onPress"]
  onSuccess?: (value: any) => void
}

export function PermissionsModal({ body, onClose, onSuccess }: PermissionsModalProps) {
  const {
    $modalBodyContainer,
    $title,
    $buttonsContainer,
    $skipButton,
    $skipButtonText,
    $nextButton,
    $nextButtonText,
    $close,
  } = useStyles(styles)

  return (
    <View
      accessible
      accessibilityLabel="coach mark"
      accessibilityHint="coach mark"
      accessibilityRole="text"
      style={$modalBodyContainer}
    >
      <Icon
        icon="x"
        accessible
        accessibilityLabel="x button"
        accessibilityHint="close modal"
        accessibilityRole="button"
        color={colors.palette.neutral450}
        onPress={onClose}
        containerStyle={$close}
        size={36}
      />
      <Text text={body} style={$title} />
      <View style={$buttonsContainer}>
        <Button
          accessible
          accessibilityLabel="close button"
          accessibilityHint="close coach mark"
          tx="permissionsModal.close"
          textStyle={$skipButtonText}
          style={$skipButton}
          pressedStyle={$skipButton}
          onPress={onClose}
        />
        <Button
          accessible
          accessibilityLabel="open Settings button"
          accessibilityHint="open Settings coach mark"
          tx="permissionsModal.openSettings"
          textStyle={$nextButtonText}
          style={$nextButton}
          pressedStyle={$nextButton}
          onPress={onSuccess}
        />
      </View>
    </View>
  )
}

const styles: StyleFn = ({ scale, fontSizes, lineHeights }) => {
  const $modalBodyContainer: ViewStyle = {
    backgroundColor: colors.palette.buttonBlue,
    borderRadius: scale(16),
    alignItems: "center",
    paddingVertical: 36,
    paddingHorizontal: 30,
    width: "100%",
    alignSelf: "center",
    marginTop: normalizeHeight(0.28),
  }

  const $title: TextStyle = {
    fontFamily: typography.primary.normal,
    fontSize: fontSizes[24],
    lineHeight: lineHeights[29],
    color: colors.palette.neutral100,
    paddingBottom: 12,
    paddingTop: 18,
  }

  const $buttonsContainer: ViewStyle = {
    flexDirection: "row",
    justifyContent: "space-between",
    width: "100%",
    alignSelf: "flex-end",
  }

  const $skipButton: ViewStyle = {
    backgroundColor: "transparent",
    borderWidth: 0,
    height: scale(56),
    minWidth: scale(140),
  }

  const $skipButtonText: TextStyle = {
    fontFamily: typography.primary.medium,
    fontSize: fontSizes[18],
    lineHeight: lineHeights[22],
    color: colors.palette.neutral100,
  }

  const $nextButton: ViewStyle = {
    height: scale(56),
    backgroundColor: colors.palette.neutral100,
    borderRadius: scale(28),
    borderWidth: 0,
    minWidth: scale(140),
  }

  const $nextButtonText: TextStyle = {
    fontFamily: typography.primary.medium,
    fontSize: fontSizes[18],
    lineHeight: lineHeights[22],
    color: colors.palette.buttonBlue,
  }

  const $close: ViewStyle = {
    position: "absolute",
    top: 0,
    right: 0,
    padding: scale(18),
    zIndex: 5,
  }

  return {
    $modalBodyContainer,
    $title,
    $buttonsContainer,
    $skipButton,
    $skipButtonText,
    $nextButton,
    $nextButtonText,
    $close,
  }
}
