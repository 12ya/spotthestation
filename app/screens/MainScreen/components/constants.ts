export const GLOBE_RADIUS = 300
export const GLOBE_SEGMENTS = 300
