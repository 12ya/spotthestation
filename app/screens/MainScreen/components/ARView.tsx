import React, { useCallback, useEffect, useState, forwardRef, useMemo } from "react"
import { Image, ImageStyle, View, ViewStyle } from "react-native"
import { ViroARSceneNavigator } from "@viro-community/react-viro"

import { colors } from "../../../theme"
import { Compass } from "./Compass"
import { DirectionCircle } from "./DirectionCircle"
import { RecordingIndicator } from "./RecordingIndicator"
import { ISSSceneAR, emitter } from "./ISSSceneAR"
import { azAltToCartesian, cartesianToAzAlt, normalizeHeading } from "../../../utils/geometry"
import { OrbitPoint } from "../../../services/api"
import { CatmullRomCurve3, Vector3 } from "three"
import { StyleFn, useStyles } from "../../../utils/useStyles"

interface ARViewProps {
  isFullScreen: boolean
  isPathVisible: boolean
  isRecording: boolean
  recordedSeconds: number
  issPath: OrbitPoint[]
  setIsSpotted: (value: boolean) => void
  image?: string
  onImageLoaded?: () => void
}

export const ARView = forwardRef<ViroARSceneNavigator, ARViewProps>(function ARView(
  { isFullScreen, isPathVisible, isRecording, recordedSeconds, issPath, setIsSpotted, image, onImageLoaded },
  ref,
) {
  const { $container, $hudContainer, $image } = useStyles(styles)

  const [issPosition, setIssPosition] = useState<[number, number]>(null)
  const [curve, setCurve] = useState<CatmullRomCurve3>()
  const [curveStartsAt, setCurveStartsAt] = useState(0)
  const [curveEndsAt, setCurveEndsAt] = useState(0)
  const [position, setPosition] = useState([0, 0])
  const onScreenPositionChange = useCallback((pos: [number, number]) => {
    setPosition(pos)
  }, [])

  const viroAppProps = useMemo(
    () => ({
      onScreenPositionChange,
    }),
    [onScreenPositionChange],
  )

  const intialScene = useMemo(
    () => ({
      scene: ISSSceneAR as any,
    }),
    [onScreenPositionChange],
  )

  useEffect(() => {
    emitter.emit("settings", { isPathVisible })
  }, [isPathVisible])

  useEffect(() => {
    if (issPath.length === 0) {
      return
    }

    setCurve(
      new CatmullRomCurve3(
        issPath.map(
          (p) => new Vector3(...azAltToCartesian(normalizeHeading(p.azimuth), p.elevation, 10)),
        ),
      ),
    )

    setCurveStartsAt(new Date(issPath[0].date).valueOf())
    setCurveEndsAt(new Date(issPath[issPath.length - 1].date).valueOf())
  }, [issPath])

  useEffect(() => {
    emitter.emit("issData", { curve, curveStartsAt, curveEndsAt })
  }, [curve])

  useEffect(() => {
    if (!curve) {
      setIssPosition(null)
      return undefined
    }
    const update = () => {
      const t =
        (Date.now() - new Date(issPath[0].date).valueOf()) /
        (new Date(issPath[issPath.length - 1].date).valueOf() - new Date(issPath[0].date).valueOf())

      let point: Vector3
      try {
        point = curve.getPoint(t)
      } catch (e) {
        console.error(e)
        return
      }

      setIssPosition(cartesianToAzAlt([point.x, point.y, point.z]))
    }
    update()

    const timeout = setInterval(update, 10000)
    return () => {
      clearInterval(timeout)
    }
  }, [curve])

  return (
    <View style={$container}>
      <ViroARSceneNavigator
        ref={ref}
        worldAlignment="GravityAndHeading"
        autofocus={true}
        initialScene={intialScene}
        viroAppProps={viroAppProps}
        style={$container}
      />

      { Boolean(image) && (
        <Image
          source={{ uri: image }}
          style={$image as ImageStyle}
          onLoad={onImageLoaded}
        />)}

      {isFullScreen && (
        <DirectionCircle screenX={position[0]} screenY={position[1]} setIsSpotted={setIsSpotted} />
      )}

      <View style={$hudContainer}>
        {Boolean(issPosition) && (
          <Compass issPosition={normalizeHeading(issPosition[0])} isFullScreen={isFullScreen} />
        )}
        {isRecording && <RecordingIndicator recordedSeconds={recordedSeconds} />}
      </View>
    </View>
  )
})

const styles: StyleFn = () => {
  const $container: ViewStyle = {
    flex: 1,
    backgroundColor: colors.backgroundDark,
  }

  const $hudContainer: ViewStyle = {
    position: "absolute",
    alignItems: "center",
    width: "100%",
    top: 0,
  }

  const $image: ViewStyle = {
    width: '100%',
    height: '100%',
    position: 'absolute'
  }

  return { $container, $hudContainer, $image }
}
