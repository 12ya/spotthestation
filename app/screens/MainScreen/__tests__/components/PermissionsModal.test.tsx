/* eslint-disable @typescript-eslint/no-unsafe-call */
import React from "react"
import { PermissionsModal } from "../../components/PermissionsModal"
import { render } from "@testing-library/react-native"

describe("PermissionsModal", () => {
  it("renders correctly", () => {
    const component = render(
      <PermissionsModal body="Test body" onClose={jest.fn()} onSuccess={jest.fn()} />,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("calls onClose callback when close button is pressed", () => {
    const onCloseMock = jest.fn()
    const component = render(
      <PermissionsModal body="Test body" onClose={onCloseMock} onSuccess={jest.fn()} />,
    )
    const closeButton = component.root.findByProps({ accessibilityLabel: "x button" })
    closeButton.props.onPress()
    expect(onCloseMock).toHaveBeenCalled()
  })

  it("calls onSuccess callback when open Settings button is pressed", () => {
    const onSuccessMock = jest.fn()
    const component = render(
      <PermissionsModal body="Test body" onClose={jest.fn()} onSuccess={onSuccessMock} />,
    )
    const openSettingsButton = component.root.findByProps({
      accessibilityLabel: "open Settings button",
    })
    openSettingsButton.props.onPress()
    expect(onSuccessMock).toHaveBeenCalled()
  })
})
