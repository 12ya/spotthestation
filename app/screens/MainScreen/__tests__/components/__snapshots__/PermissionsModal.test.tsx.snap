// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsModal renders correctly 1`] = `
<View
  accessibilityHint="coach mark"
  accessibilityLabel="coach mark"
  accessibilityRole="text"
  accessible={true}
  style={
    {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#0B3D91",
      "borderRadius": 23,
      "marginTop": 373.52000000000004,
      "paddingHorizontal": 30,
      "paddingVertical": 36,
      "width": "100%",
    }
  }
>
  <View
    accessibilityHint="close modal"
    accessibilityLabel="x button"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "padding": 26,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 5,
      }
    }
  >
    <RCTImageView
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      internal_analyticTag={null}
      resizeMode="contain"
      source={
        [
          {
            "height": undefined,
            "uri": undefined,
            "width": undefined,
          },
        ]
      }
      style={
        {
          "height": 52,
          "overflow": "hidden",
          "resizeMode": "contain",
          "tintColor": "#ADADAE",
          "width": 52,
        }
      }
      tintColor="#ADADAE"
    />
  </View>
  <Text
    allowFontScaling={false}
    style={
      [
        {},
        [
          {
            "fontSize": 23,
            "lineHeight": 35,
          },
          {
            "fontFamily": "Helvetica Neue",
          },
          {
            "color": "#191015",
          },
        ],
        undefined,
        undefined,
        {
          "color": "#FFFFFF",
          "fontFamily": "Helvetica Neue",
          "fontSize": 35,
          "lineHeight": 42,
          "paddingBottom": 12,
          "paddingTop": 18,
        },
      ]
    }
  >
    Test body
  </Text>
  <View
    style={
      {
        "alignSelf": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <View
      accessibilityHint="close coach mark"
      accessibilityLabel="close button"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          [
            {
              "alignItems": "center",
              "borderRadius": 6,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 80,
              "overflow": "hidden",
              "paddingHorizontal": 17,
              "paddingVertical": 17,
            },
            {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#B6ACA6",
              "borderWidth": 1,
            },
          ],
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
            "height": 80,
            "minWidth": 201,
          },
          false,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          [
            {},
            [
              {
                "fontSize": 23,
                "lineHeight": 35,
              },
              {
                "fontFamily": "Helvetica Neue",
              },
              {
                "color": "#191015",
              },
            ],
            undefined,
            undefined,
            [
              {
                "flexGrow": 0,
                "flexShrink": 1,
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 23,
                "lineHeight": 29,
                "textAlign": "center",
                "zIndex": 2,
              },
              {
                "color": "#FFFFFF",
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 26,
                "lineHeight": 32,
              },
              false,
            ],
          ]
        }
      >
        permissionsModal.close undefined
      </Text>
    </View>
    <View
      accessibilityHint="open Settings coach mark"
      accessibilityLabel="open Settings button"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          [
            {
              "alignItems": "center",
              "borderRadius": 6,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 80,
              "overflow": "hidden",
              "paddingHorizontal": 17,
              "paddingVertical": 17,
            },
            {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#B6ACA6",
              "borderWidth": 1,
            },
          ],
          {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 40,
            "borderWidth": 0,
            "height": 80,
            "minWidth": 201,
          },
          false,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        style={
          [
            {},
            [
              {
                "fontSize": 23,
                "lineHeight": 35,
              },
              {
                "fontFamily": "Helvetica Neue",
              },
              {
                "color": "#191015",
              },
            ],
            undefined,
            undefined,
            [
              {
                "flexGrow": 0,
                "flexShrink": 1,
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 23,
                "lineHeight": 29,
                "textAlign": "center",
                "zIndex": 2,
              },
              {
                "color": "#0B3D91",
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 26,
                "lineHeight": 32,
              },
              false,
            ],
          ]
        }
      >
        permissionsModal.openSettings undefined
      </Text>
    </View>
  </View>
</View>
`;
